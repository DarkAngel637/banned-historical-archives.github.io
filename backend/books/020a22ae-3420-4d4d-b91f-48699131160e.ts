export default {
  entity: {
    id: '020a22ae-3420-4d4d-b91f-48699131160e',
    name: '邓小平与孔孟之道（初编）',
    internal: false,
    official: false,
    type: 'img',
    author: '',
    files: new Array(13)
      .fill(0)
      .map(
        (i, idx) =>
          `https://raw.githubusercontent.com/banned-historical-archives/banned-historical-archives1/main/020a22ae-3420-4d4d-b91f-48699131160e/${
            idx + 1
          }.`,
      )
      .join(','),
  },
  parser_option: {
    page_limits: [],
    ext: '',
    articles: [
      {
        title: '邓小平与孔孟之道（初编）',
        authors: ['北京大学清华大学材料组'],
        page_start: 1,
        page_end: 13,
        dates: [{ year: 1976, month: 5 }],
      },
    ],
    ocr: undefined,
    ocr_exceptions: {},
  },
  parser_id: 'automation',
  path: '/archives1/020a22ae-3420-4d4d-b91f-48699131160e',
};
