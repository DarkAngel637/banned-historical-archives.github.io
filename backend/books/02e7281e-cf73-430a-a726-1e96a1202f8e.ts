export default {
  entity: {
    id: '02e7281e-cf73-430a-a726-1e96a1202f8e',
    name: '六月八日下午军区25次党委扩大会全体会议李德生同志的讲话',
    internal: false,
    official: false,
    type: 'img',
    author: '',
    files: new Array(16)
      .fill(0)
      .map(
        (i, idx) =>
          `https://raw.githubusercontent.com/banned-historical-archives/banned-historical-archives16/main/02e7281e-cf73-430a-a726-1e96a1202f8e/${
            idx + 1
          }.`,
      )
      .join(','),
  },
  parser_option: {
    page_limits: [],
    ext: '',
    articles: [
      {
        title: '六月八日下午军区25次党委扩大会全体会议李德生同志的讲话',
        authors: ['李德生'],
        page_start: 1,
        page_end: 16,
        dates: [{ year: 1973, month: 6, day: 8 }],
      },
    ],
    ocr: undefined,
    ocr_exceptions: {},
  },
  parser_id: 'automation',
  path: '/archives16/02e7281e-cf73-430a-a726-1e96a1202f8e',
};
