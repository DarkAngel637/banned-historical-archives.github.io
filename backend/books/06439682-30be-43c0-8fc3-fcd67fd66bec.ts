
export default {"entity":{"id":"06439682-30be-43c0-8fc3-fcd67fd66bec","name":"我为什么参加红色造反团","internal":true,"official":true,"type":"img","author":"","files":"https://raw.githubusercontent.com/banned-historical-archives/banned-historical-archives1/main/06439682-30be-43c0-8fc3-fcd67fd66bec/1.jpg,https://raw.githubusercontent.com/banned-historical-archives/banned-historical-archives1/main/06439682-30be-43c0-8fc3-fcd67fd66bec/2.jpg"},"parser_option":{"page_limits":[],"ext":"jpg","articles":[{"title":"我为什么参加红色造反团","authors":["毛远新"],"page_start":1,"page_end":2,"dates":[{"year":1966,"month":12,"day":11}]}],"ocr_exceptions":{}},"parser_id":"automation","path":"/archives1/06439682-30be-43c0-8fc3-fcd67fd66bec"};
