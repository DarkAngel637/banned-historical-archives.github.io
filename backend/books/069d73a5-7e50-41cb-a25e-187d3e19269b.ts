export default {
  entity: {
    id: '069d73a5-7e50-41cb-a25e-187d3e19269b',
    name: '无产阶级文化大革命资料汇编（第二篇）（上）（3）（水陆洲2011年）',
    internal: false,
    official: false,
    type: 'pdf',
    author: '水陆洲编',
    files:
      'https://raw.githubusercontent.com/banned-historical-archives/banned-historical-archives3/main/069d73a5-7e50-41cb-a25e-187d3e19269b.pdf',
  },
  parser_option: {
    page_limits: [],
    ext: 'pdf',
    articles: [
      {
        title:
          '第二章 炮打资产阶级司令部 批判资产阶级反动路线 第八节 部队院校文革运动的开展',
        authors: ['水陆洲'],
        page_start: 1,
        page_end: 94,
        dates: [{ year: 2011, month: 8 }],
      },
      {
        title: '第三章 一九六七年一月的革命风暴 第一节 一月革命风暴',
        authors: ['水陆洲'],
        page_start: 95,
        page_end: 259,
        dates: [{ year: 2011, month: 8, day: 1 }],
      },
      {
        title: '第三章 一九六七年一月的革命风暴 第二节 上海市革命委员会成立',
        authors: ['水陆洲'],
        page_start: 260,
        page_end: 798,
        dates: [{ year: 2011, month: 8, day: 4 }],
      },
      {
        title: '第三章 一九六七年一月的革命风暴 第三节 山西省革命委员会成立',
        authors: ['水陆洲'],
        page_start: 799,
        page_end: 944,
        dates: [{ year: 2011, month: 8, day: 10 }],
      },
      {
        title: '第三章 一九六七年一月的革命风暴 第四节 贵州省革命委员会成立',
        authors: ['水陆洲'],
        page_start: 945,
        page_end: 975,
        dates: [{ year: 2011, month: 8, day: 8 }],
      },
      {
        title: '第三章 一九六七年一月的革命风暴 第五节 黑龙江省革命委员会成立',
        authors: ['水陆洲'],
        page_start: 976,
        page_end: 1090,
        dates: [{ year: 2011, month: 8, day: 9 }],
      },
      {
        title: '第三章 一九六七年一月的革命风暴 第六节 山东省革命委员会成立',
        authors: ['水陆洲'],
        page_start: 1091,
        page_end: 1241,
        dates: [{ year: 2011, month: 8, day: 11 }],
      },
      {
        title: '第三章 一九六七年一月的革命风暴 第七节 北京市革命委员会成立',
        authors: ['水陆洲'],
        page_start: 1242,
        page_end: 1723,
        dates: [{ year: 2011, month: 8, day: 13 }],
      },
      {
        title: '第三章 一九六七年一月的革命风暴 第八节 解决陶铸问题',
        authors: ['水陆洲'],
        page_start: 1724,
        page_end: 1986,
        dates: [{ year: 2011, month: 8, day: 15 }],
      },
      {
        title: '第三章 一九六七年一月的革命风暴 第九节 批刘保帅保肖',
        authors: ['水陆洲'],
        page_start: 1987,
        page_end: 2149,
        dates: [{ year: 2011, month: 8, day: 16 }],
      },
    ],
    ocr: {
      standard_paragraph_merge_strategy_threshold: 0.186,
      differential_paragraph_merge_strategy_threshold: 0,
      content_thresholds: [0.054, 0, 0, 0],
    },
    ocr_exceptions: {},
  },
  parser_id: 'automation',
  path: '/archives3/069d73a5-7e50-41cb-a25e-187d3e19269b.pdf',
};
