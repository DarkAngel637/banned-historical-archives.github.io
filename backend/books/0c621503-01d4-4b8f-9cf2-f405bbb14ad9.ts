export default {
  entity: {
    id: '0c621503-01d4-4b8f-9cf2-f405bbb14ad9',
    name: '纪念中国共产党成立一百周年（2021.7.1）',
    internal: false,
    official: false,
    type: 'pdf',
    author: '中国共产主义小组（毛主义），2021.7.1',
    files:
      'https://raw.githubusercontent.com/banned-historical-archives/banned-historical-archives3/main/0c621503-01d4-4b8f-9cf2-f405bbb14ad9.pdf',
  },
  parser_option: {
    page_limits: [],
    ext: 'pdf',
    articles: [
      {
        title: '纪念中国共产党成立一百周年',
        alias: '共产主义小组（毛主义）：纪念中国共产党成立一百周年',
        authors: ['共产主义小组（毛主义）'],
        page_start: 1,
        page_end: 23,
        tags: [
          { name: '共产主义小组（毛主义）', type: 'issuer' }, // 发行方/出版方
          { name: '新民主主义革命', type: 'subject' }, // 事件
          { name: '世界形势', type: 'subject' }, // 事件
          { name: '人民战争', type: 'subject' }, // 事件
          { name: '革命的三联合', type: 'subject' }, // 事件
          { name: '革命大联合', type: 'subject' }, // 事件
          { name: '七二〇事件', type: 'subject' }, // 事件
          { name: '党内资产阶级', type: 'subject' }, // 事件
          { name: '资本主义复辟', type: 'subject' }, // 事件
          { name: '佳士运动', type: 'subject' }, // 事件
          { name: '毛主义', type: 'subject' }, // 事件
          { name: '三大工具的同心建设', type: 'subject' }, // 事件
          { name: '两条路线斗争', type: 'subject' }, // 事件
      ],
        dates: [{ year: 2021, month: 7, day: 1 }],
      },
    ],
    ocr: {
      standard_paragraph_merge_strategy_threshold: 0.186,
      differential_paragraph_merge_strategy_threshold: 0,
    },
    ocr_exceptions: {},
  },
  parser_id: 'automation',
  path: '/archives3/0c621503-01d4-4b8f-9cf2-f405bbb14ad9.pdf',
};
