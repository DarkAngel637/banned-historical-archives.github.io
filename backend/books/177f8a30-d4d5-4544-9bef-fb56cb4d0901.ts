export default {
  entity: {
    id: '177f8a30-d4d5-4544-9bef-fb56cb4d0901',
    name: '无产阶级文化大革命资料汇编（第二篇）（上）（2）（水陆洲2011年）',
    internal: false,
    official: false,
    type: 'pdf',
    author: '水陆洲编',
    files:
      'https://raw.githubusercontent.com/banned-historical-archives/banned-historical-archives3/main/177f8a30-d4d5-4544-9bef-fb56cb4d0901.pdf',
  },
  parser_option: {
    page_limits: [],
    ext: 'pdf',
    articles: [
      {
        title:
          '第二章 炮打资产阶级司令部 批判资产阶级反动路线 第五节 红卫兵运动风起云涌和学校文革运动的开展',
        authors: ['水陆洲'],
        page_start: 1,
        page_end: 916,
        dates: [{ year: 2011, month: 7, day: 27 }],
      },
      {
        title:
          '第二章 炮打资产阶级司令部 批判资产阶级反动路线 第六节 中央机关文革运动的开展',
        authors: ['水陆洲'],
        page_start: 917,
        page_end: 1282,
        dates: [{ year: 2011, month: 7, day: 28 }],
      },
      {
        title:
          '第二章 炮打资产阶级司令部 批判资产阶级反动路线 第七节 各省市区文革运动的开展',
        authors: ['水陆洲'],
        page_start: 1283,
        page_end: 1886,
        dates: [{ year: 2011, month: 8, day: 4 }],
      },
    ],
    ocr: {
      standard_paragraph_merge_strategy_threshold: 0.186,
      differential_paragraph_merge_strategy_threshold: 0,
      content_thresholds: [0.054, 0, 0, 0],
    },
    ocr_exceptions: {},
  },
  parser_id: 'automation',
  path: '/archives3/177f8a30-d4d5-4544-9bef-fb56cb4d0901.pdf',
};
