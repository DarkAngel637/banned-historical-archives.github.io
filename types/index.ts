import type Publication from '../backend/entity/publication';

export enum ArticleCategory {
  centralFile = '中央文件',
  editorial = '重要报刊和社论',
  keyFigures = '关键人物文稿',
  keyPapersFromTheMasses = '群众运动重要文献',
}

export enum TagType {
  articleCategory = '文稿大类',
  articleType = '文稿类型',
  place = '地点',
  character = '人物',
  issuer = '出版方/发行方',
  subject = '主题/事件',
  recorder = '记录',
  reviewer = '审核',
  translator = '翻译',
  reprint = '翻印/传抄',
}

export enum ImageTagType {
  place = 'place',
  character = 'character',
  subject = 'subject',
}

export enum ContentType {
  appellation = 'appellation',
  title = 'title',
  subtitle = 'subtitle',
  subtitle2 = 'subtitle2',
  subtitle3 = 'subtitle3',
  subtitle4 = 'subtitle4',
  subtitle5 = 'subtitle5',
  subdate = 'subdate',
  paragraph = 'paragraph',
  quotation = 'quotation',
  signature = 'signature',
  image = 'image',
  image_description = 'image_description',
}

export type ContentPartRaw = {
  text: string;
  type: ContentType;
};
export type ContentPart = {
  text: string;
  type: ContentType;
};

export type Date = {
  year?: number;
  month?: number;
  day?: number;
};

export type Pivot = {
  part_idx: number; // 从 0 开始
  index: number; // 注释编号
  offset: number; // 偏移量，从 0 开始，注释应该插入的index，比如'mzd[2]'的offset为3
};

export type ParserResult = {
  title: string;
  alias?: string;
  dates: Date[];
  is_range_date: boolean;
  authors: string[];
  parts: ContentPart[];
  comments: string[];
  comment_pivots: Pivot[];
  description: string;
  page_start: number;
  page_end: number;
  origin?: string; // 起源
  tags?: {
    name: string;
    type: TagType;
  }[];
  file_id?: string;

  title_raw?: string;
  date_raw?: string;
  parts_raw?: ContentPartRaw[];
};

export enum ArticleType {
  writings = '文章',
  mail = '书信',
  lecture = '发言',
  talk = '对话',
  declaration = '宣言',
  instruction = '指示',
  comment = '批示',
  telegram = '通讯',
}

export type OCRPosition = [number, number];
export type OCRResult = {
  // 坐标轴原点在左上角，y轴朝下
  // 左上，右上，右下，左下
  box: [OCRPosition, OCRPosition, OCRPosition, OCRPosition];
  text: string;
};
export type LACType =
  | 'n' // 普通名词
  | 'f' // 方位名词
  | 's' // 处所名词
  | 't' // 时间
  | 'nr' // 人名
  | 'ns' // 地名
  | 'nt' // 机构名
  | 'nw' // 作品名
  | 'nz' // 其他专名
  | 'v' // 普通动词
  | 'vd' // 动副词
  | 'vn' // 名动词
  | 'a' // 形容词
  | 'ad' // 副形词
  | 'an' // 名形词
  | 'd' // 副词
  | 'm' // 数量词
  | 'q' // 量词
  | 'r' // 代词
  | 'p' // 介词
  | 'c' // 连词
  | 'u' // 助词
  | 'xc' // 其他虚词
  | 'w' // 标点符号
  | 'PER' // 人名
  | 'LOC' // 地名
  | 'ORG' // 机构名
  | 'TIME'; // 时间

export type LACResult = {
  text: string;
  count: number;
  type: LACType;
};

export type ParserOptionV2 = {
  archive_id?: number;
  internal?: boolean;
  official?: boolean;
  author?: string;
  articles?: {
    title: string;
    authors: string[]; // 作者
    dates: Date[]; // 时间 或者 时间范围 或者 多个时间点
    is_range_date?: boolean; // 如果为 true 表示一段时间，如果为false表示多/单个时间点
    alias?: string; // 标题别名
    ocr?: Partial<OCRParameter & OCRParameterAdvanced>; // 此参数比全局ocr参数的优先级高，默认为空
    ocr_exceptions?: {
      [key: string]: Partial<OCRParameter & OCRParameterAdvanced>;
    };
    tags?: { name: string; type: keyof typeof TagType }[];
    page_start: number;
    page_end: number;
  }[];
  ext?: string;
  ocr?: Partial<OCRParameter & OCRParameterAdvanced>; // ocr 全局参数
  ocr_exceptions?: {
    [key: string]: Partial<OCRParameter & OCRParameterAdvanced>;
  }; // 例外， 比如第三页的ocr参数与其他页面不同，默认为空，此参数比articles中的优先级高
};

export type ParserOption = ParserOptionV2 & {
  page_limits: [number, number][];

  /*legacy*/
  page_width?: number;
  content_min_x?: number;
  name?: string;
  header_min_height?: number;
};

export type OCRParameter = {
  rec_model: string;
  rec_backend: string;
  det_model: string;
  det_backend: string;
  resized_shape: number;
  box_score_thresh: number;
  min_box_size: number;
};

export type OCRParameterAdvanced = {
  line_merge_threshold: number; // 单位像素，如果小于这个阈值将被视为同一行
  standard_paragraph_merge_strategy_threshold: number; // 标准段落合并策略，超过 threshold * width 的表示新段落，否则向上合并
  differential_paragraph_merge_strategy_threshold: number; // 差分段落合并策略，x[i] - x[i-1] > threshold 的表示新段落，否则向上合并，单位像素
  content_thresholds: [number, number, number, number]; // 通常需要忽略在页面边缘的页眉，页码或者噪声，数组内4个数值分别表示上下左右相对于宽高的比例， 例如 [0.1,0,0,0] 表示忽略顶部占总高度百分之10的内容
  auto_vsplit: boolean; // 用于分页或者处理特殊的排版。如果为 ture，当页面宽度大于高度时，将ocr结果中页面中间(vsplit的位置)分开
  vsplit: number; // 如果设置为0.5，ocr结果将从页面宽度的50%处分割，如果为0表示不分割。当auto_vsplit为false且vsplit不为0时，表示任何页面都进行分割。
};

export type Book = {
  entity: Partial<Publication>;
  path: string;
  parser_option: ParserOption;
  parser_id: string;
  parser: (path: string, opt: ParserOption) => Promise<ParserResult[]>;
};

export type Patch = {
  parts: { [idx: string]: string };
  comments: { [idx: string]: string };
  description: string;
};

export type StringDiff = string;

export type PartDiff = {
  insertBefore?: ContentPart[];
  insertAfter?: ContentPart[];
  delete?: boolean;
  diff?: StringDiff;
  type?: ContentType;
};

export type CommentDiff = {
  insertBefore?: { id?: string; text: StringDiff }[];
  insertAfter?: { id?: string; text: StringDiff }[];
  delete?: boolean;
  diff?: StringDiff;
};

export type PatchV2 = {
  version: 2;
  parts: { [idx: string]: PartDiff };
  comments: { [idx: string]: CommentDiff };
  newComments?: string[];
  description?: StringDiff; // 如果为空字符串表示无变更，如果不存在，表示删除
};
