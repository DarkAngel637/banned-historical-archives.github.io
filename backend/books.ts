import {join} from 'node:path';
import {Book, ParserOption} from '../types';
import * as jinghuo_parser from './parser/jinghuo_parser';
import * as wansui_parser from './parser/wansui_parser';
import * as wenji_parser from './parser/wenji_parser';
import * as jqjianghua_parser from './parser/jqjianghua_parser';
import * as xuanji_parser from './parser/xuanji';

const books: Book[] = [
  {
    entity: {
      id: 'maoxuan-jinghuo',
      name: '毛泽东选集1-7（静火版）',
      internal: false,
      official: false,
      author: '静火',
      pdf: '/books/maoxuan-jinghuo.pdf',
    },
    parser_option: {
      page_limits: [
        [20, 1183],
        [1192, 1489],
        [1504, 1791],
      ],
    },
    parser: jinghuo_parser.parse,
    path: join(__dirname, '../public/books/maoxuan-jinghuo.pdf'),
  },
  {
    entity: {
      id: 'wansui1',
      name: '毛泽东思想万岁（1913-1943卷）',
      internal: false,
      official: false,
      author: '未知',
      pdf: '/books/wansui1a.pdf',
    },
    parser_option: {
      page_limits: [[9, 569]],
    },
    parser: wansui_parser.parse,
    path: join(__dirname, '../public/books/wansui1a.pdf'),
  },
  {
    entity: {
      id: 'wansui2',
      name: '毛泽东思想万岁（1943-1949卷）',
      internal: false,
      official: false,
      author: '未知',
      pdf: '/books/wansui2a.pdf',
    },
    parser_option: {
      page_limits: [[9, 542]],
    },
    parser: wansui_parser.parse,
    path: join(__dirname, '../public/books/wansui2a.pdf'),
  },
  {
    entity: {
      id: 'wansui3',
      name: '毛泽东思想万岁（1949-1957卷）',
      internal: false,
      official: false,
      author: '未知',
      pdf: '/books/wansui3a.pdf',
    },
    parser_option: {
      page_limits: [[9, 508]],
    },
    parser: wansui_parser.parse,
    path: join(__dirname, '../public/books/wansui3a.pdf'),
  },
  {
    entity: {
      id: 'wansui4',
      name: '毛泽东思想万岁（1958-1960卷）',
      internal: false,
      official: false,
      author: '未知',
      pdf: '/books/wansui4a.pdf',
    },
    parser_option: {
      page_limits: [[11, 674]],
    },
    parser: wansui_parser.parse,
    path: join(__dirname, '../public/books/wansui4a.pdf'),
  },
  {
    entity: {
      id: 'wansui5',
      name: '毛泽东思想万岁（1961-1968卷）',
      internal: false,
      official: false,
      author: '未知',
      pdf: '/books/wansui5a.pdf',
    },
    parser_option: {
      page_limits: [[16, 668]],
    },
    parser: wansui_parser.parse,
    path: join(__dirname, '../public/books/wansui5a.pdf'),
  },
  {
    entity: {
      id: 'wenji1',
      name: '毛泽东文集第一卷',
      internal: false,
      official: true,
      author: '中央文献研究室',
      pdf: '/books/wenji1.pdf',
    },
    parser_option: {
      page_limits: [[14, 526]],
      page_width: 435.2,
      content_min_x: 74.5,
    },
    parser: wenji_parser.parse,
    path: join(__dirname, '../public/books/wenji1.pdf'),
  },
  {
    entity: {
      id: 'wenji2',
      name: '毛泽东文集第二卷',
      internal: false,
      official: true,
      author: '中央文献研究室',
      pdf: '/books/wenji2.pdf',
    },
    parser_option: {
      page_limits: [[14, 481]],
      page_width: 393.2,
      content_min_x: 44.88,
    },
    parser: wenji_parser.parse,
    path: join(__dirname, '../public/books/wenji2.pdf'),
  },
  {
    entity: {
      id: 'wenji3',
      name: '毛泽东文集第三卷',
      internal: false,
      official: true,
      author: '中央文献研究室',
      pdf: '/books/wenji3.pdf',
    },
    parser_option: {
      page_limits: [[12, 467]],
      page_width: 442,
      content_min_x: 74.28,
    },
    parser: wenji_parser.parse,
    path: join(__dirname, '../public/books/wenji3.pdf'),
  },
  {
    entity: {
      id: 'wenji4',
      name: '毛泽东文集第四卷',
      internal: false,
      official: true,
      author: '中央文献研究室',
      pdf: '/books/wenji4.pdf',
    },
    parser_option: {
      page_limits: [[15, 354]],
      page_width: 442,
      content_min_x: 74.28,
    },
    parser: wenji_parser.parse,
    path: join(__dirname, '../public/books/wenji4.pdf'),
  },
  {
    entity: {
      id: 'wenji5',
      name: '毛泽东文集第五卷',
      internal: false,
      official: true,
      author: '中央文献研究室',
      pdf: '/books/wenji5.pdf',
    },
    parser_option: {
      page_limits: [[14, 363]],
      page_width: 382,
      content_min_x: 45.84,
    },
    parser: wenji_parser.parse,
    path: join(__dirname, '../public/books/wenji5.pdf'),
  },
  {
    entity: {
      id: 'wenji6',
      name: '毛泽东文集第六卷',
      internal: false,
      official: true,
      author: '中央文献研究室',
      pdf: '/books/wenji6.pdf',
    },
    parser_option: {
      page_limits: [[15, 527]],
      page_width: 449.5,
      content_min_x: 72.28,
    },
    parser: wenji_parser.parse,
    path: join(__dirname, '../public/books/wenji6.pdf'),
  },
  {
    entity: {
      id: 'wenji7',
      name: '毛泽东文集第七卷',
      internal: false,
      official: true,
      author: '中央文献研究室',
      pdf: '/books/wenji7.pdf',
    },
    parser_option: {
      page_limits: [[10, 474]],
      page_width: 389.5,
      content_min_x: 43.24,
    },
    parser: wenji_parser.parse,
    path: join(__dirname, '../public/books/wenji7.pdf'),
  },
  {
    entity: {
      id: 'wenji8',
      name: '毛泽东文集第八卷',
      internal: false,
      official: true,
      author: '中央文献研究室',
      pdf: '/books/wenji8.pdf',
    },
    parser_option: {
      page_limits: [[12, 457]],
      page_width: 389.5,
      content_min_x: 43.24,
    },
    parser: wenji_parser.parse,
    path: join(__dirname, '../public/books/wenji8.pdf'),
  },
  {
    entity: {
      id: 'jqjianghua',
      name: '江青十年讲话汇编',
      internal: false,
      official: false,
      author: '',
      pdf: '/books/jqjianghua.pdf',
    },
    parser_option: {
      page_limits: [[7, 730], [736, 752]],
    },
    parser: jqjianghua_parser.parse,
    path: join(__dirname, '../public/books/jqjianghua.pdf'),
  },
  {
    entity: {
      id: 'xuanji1',
      name: '毛泽东选集第一卷(1967)',
      internal: false,
      official: true,
      author: '毛泽东',
      pdf: '/books/毛泽东选集(1967)1.pdf',
    },
    parser_option: {
      page_limits: [[12, 321]],
      name: 'xuanji1',
    },
    parser: xuanji_parser.parse,
    path: join(__dirname, '../public/books/毛泽东选集(1967)1.pdf'),
  },
];
export default books;