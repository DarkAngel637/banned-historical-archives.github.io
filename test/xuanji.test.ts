import { assert, expect, test, it, describe, beforeEach, beforeAll } from 'vitest';
import { join } from 'node:path';
import books, { get_book } from '../backend/books';
import { ContentType, ParserResult } from '../types';

function n_comments_equals_n_pivots(res: ParserResult[]) {
  for (const i of res) {
    const temp: { [key: string]: boolean } = {};
    i.comment_pivots.forEach((j) => {
      temp[j.index] = true;
    });
    const matched = Object.keys(temp).length === i.comments.length;
    if (!matched) return false;
  }
  return true;
}

describe('xuanji1', async () => {
  const book = await get_book('xuanji1')!;
  const res = await book.parser(book.path, book.parser_option);
  it('title & dates', () => {
    expect(res.map((i) => i.title)).toMatchSnapshot();
    expect(res.map((i) => i.dates)).toMatchSnapshot();
  });
  it('comments & pivots', () => {
    expect(n_comments_equals_n_pivots(res)).toBe(true);
  });
  it('alignment', () => {
    expect(res[1]).toMatchSnapshot();
  });
});

describe('xuanji2', async () => {
  const book = await get_book('xuanji2')!;
  const res = await book.parser(book.path, book.parser_option);
  it('title & dates', () => {
    expect(res.map((i) => i.title)).toMatchSnapshot();
    expect(res.map((i) => i.dates)).toMatchSnapshot();
  });
  it('comments & pivots', () => {
    expect(n_comments_equals_n_pivots(res)).toBe(true);
  });
});

describe('xuanji3', async () => {
  const book = await get_book('xuanji3')!;
  const res = await book.parser(book.path, book.parser_option);
  it('title & dates', () => {
    expect(res.map((i) => i.title)).toMatchSnapshot();
    expect(res.map((i) => i.dates)).toMatchSnapshot();
  });
  it('comments & pivots', () => {
    expect(n_comments_equals_n_pivots(res)).toBe(true);
  });
});

describe('xuanji4', async () => {
  const book = await get_book('xuanji4')!;
  const res = await book.parser(book.path, book.parser_option);
  it('title & dates', () => {
    expect(res.map((i) => i.title)).toMatchSnapshot();
    expect(res.map((i) => i.dates)).toMatchSnapshot();
  });
  it('comments & pivots', () => {
    expect(n_comments_equals_n_pivots(res)).toBe(true);
  });
  it('description merge', () => {
    expect(
      res.filter((i) => i.title === '第十八集团军总司令给蒋介石的两个电报'),
    ).toMatchSnapshot();
  });
});

describe('xuanji5', async () => {
  const book = await get_book('xuanji5')!;
  const res = await book.parser(book.path, book.parser_option);
  it('title & dates', () => {
    expect(res.map((i) => i.title)).toMatchSnapshot();
    expect(res.map((i) => i.dates)).toMatchSnapshot();
  });
  it('comments & pivots', () => {
    expect(n_comments_equals_n_pivots(res)).toBe(true);
  });
  it('alignment', () => {
    expect(res.filter(i => i.title === '中共中央关于西藏工作方针的指示')[0]).toMatchSnapshot();
    expect(res.filter(i => i.title === '永远保持艰苦奋斗的作风')[0]).toMatchSnapshot();
  });
});