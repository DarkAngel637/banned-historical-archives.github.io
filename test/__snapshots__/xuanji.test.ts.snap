// Vitest Snapshot v1

exports[`xuanji1 > title & dates 1`] = `
[
  "中国社会各阶级的分析",
  "湖南农民运动考察报告",
  "中国的红色政权为什么能够存在？",
  "井冈山的斗争",
  "关于纠正党内的错误思想",
  "星星之火，可以燎原",
  "必须注意经济工作",
  "怎样分析农村阶级",
  "我们的经济政策",
  "关心群众生活，注意工作方法",
  "论反对日本帝国主义的策略",
  "中国革命战争的战略问题",
  "关于蒋介石声明的声明",
  "中国共产党在抗日时期的任务",
  "为争取千百万群众进入抗日民族统一战线而斗争",
  "实践论",
  "矛盾论",
]
`;

exports[`xuanji1 > title & dates 2`] = `
[
  [
    {
      "day": 1,
      "month": 12,
      "year": 1925,
    },
  ],
  [
    {
      "day": undefined,
      "month": 3,
      "year": 1927,
    },
  ],
  [
    {
      "day": 5,
      "month": 10,
      "year": 1928,
    },
  ],
  [
    {
      "day": 25,
      "month": 11,
      "year": 1928,
    },
  ],
  [
    {
      "day": undefined,
      "month": 12,
      "year": 1929,
    },
  ],
  [
    {
      "day": 5,
      "month": 1,
      "year": 1930,
    },
  ],
  [
    {
      "day": 12,
      "month": 8,
      "year": 1933,
    },
  ],
  [
    {
      "day": undefined,
      "month": 10,
      "year": 1933,
    },
  ],
  [
    {
      "day": undefined,
      "month": 1,
      "year": 1934,
    },
  ],
  [
    {
      "day": 27,
      "month": 1,
      "year": 1934,
    },
  ],
  [
    {
      "day": 27,
      "month": 12,
      "year": 1935,
    },
  ],
  [
    {
      "day": undefined,
      "month": 12,
      "year": 1936,
    },
  ],
  [
    {
      "day": 28,
      "month": 12,
      "year": 1936,
    },
  ],
  [
    {
      "day": 3,
      "month": 5,
      "year": 1937,
    },
  ],
  [
    {
      "day": 8,
      "month": 5,
      "year": 1937,
    },
  ],
  [
    {
      "day": undefined,
      "month": 7,
      "year": 1937,
    },
  ],
  [
    {
      "day": undefined,
      "month": 8,
      "year": 1937,
    },
  ],
]
`;
